### ISSUE TITLE
Advanced mode 1 Failing on Ubuntu 20.04 / Python 3.6.9

### ISSUE THREAD

**Issue by e0lithic** [dashingsoft/pyarmor#243]
Hi Again,

## Environment
Python - 3.6.9
OS - Ubuntu 20.04
Pyarmor - 6.2.0

## Project structure
mypkg
├── bar.py
├── foo.py
└── __init__.py

## Configuration files
I am using the child project structure to demonstrate a use-case for a large package. Here we are just obfuscating everything else with restrict mode 4 and using restrict mode 1 for `__init__.py`. As is recommended for a package.
``` python
# Content .pyarmor_config
{ 
  "version": "2.0",
  "name": "mypkg",
  "title": "mypkg",
  "src": ".",
  "is_package": 1,
  "manifest": "include __init__.py",
  "entry": "__init__.py",
  "output": "dist",
  "runtime_path": null,
  "restrict_mode": 1,
  "obf_code": 2,
  "obf_mod": 1,
  "wrap_mode": 1,
  "advanced_mode": 0,
  "bootstrap_code": 1,
  "cross_protection": 1,
  "plugins": null,
  "platform": null,
  "package_runtime": 1,
  "enable_suffix": 0,
  "license_file": null,
  "build_time": 1589994512.6738808
}
```
```python
# Content .pyarmor_config_pkg
{
  "version": "2.0",
  "name": "mypkg",
  "title": "mypkg",
  "src": ".",
  "is_package": 1,
  "manifest": "global-include *.py, exclude __init__.py",
  "entry": "",
  "output": "dist",
  "runtime_path": null,
  "restrict_mode": 4,
  "obf_code": 2,
  "obf_mod": 1,
  "wrap_mode": 1,
  "advanced_mode": 0,
  "bootstrap_code": 1,
  "cross_protection": 1,
  "plugins": null,
  "platform": null,
  "package_runtime": 1,
  "enable_suffix": 0,
  "license_file": null,
  "build_time": 1589994515.6668053
}
```
**Command** :
```bash
pyarmor build -B
pyarmor build --no-runtime -B .pyarmor_config_pkg
```
The build completes successfully.

# Behaviour/Expectation
If we try to import the obfuscated mykg then we get the following error
```bash
Python 3.6.9 (default, May 20 2020, 20:21:06) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import mypkg
XXX lineno: 8, opcode: 228
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "</media/e0lithic/D/GameX/pyarmor_test/mypkg/dist/mypkg/__init__.py>", line 3, in <module>
  File "<frozen mypkg>", line 8, in <module>
SystemError: unknown opcode
```
If we just turn `advanced_mode` 0 in both configuration files and rerun the build. I am able to import the package.

I read the documentation but the python 3.6 limitation was only mentioned for advanced_mode 2 / a.k.a super mode. Is the support for advanced mode 1 on python 3.6 also depreciated? Kindly let me know if I am missing something.

Thanks

**Comment by jondy** [comment]
Does it work just obfuscating one simple script with advanced 1? It should work in Ubuntu /Python 3.6.

**Comment by e0lithic** [comment]
I tried it on a single file and the issue persists.

**Comment by jondy** [comment]
Could you upload the file `libpython3.6m.so` in your Ubuntu20.04 to "https://tmpfiles.org/" and send me a link? I need check it.

**Comment by e0lithic** [comment]
Here is the download link for the [same](https://tmpfiles.org/download/50379/libpython3.6m.so). Just for additional information I am using pyenv and instead of a clean installation, I have upgraded to Ubuntu to 20.04 from 18.04

**Comment by e0lithic** [comment]
Since you mentioned the involvemenet of libpython for the .so files. I think this could very well be the case of pyenv not being able to link the libpython library as mentioned in the [issue](https://github.com/pyenv/pyenv/issues/65). I will try enabling shared libraries and then give it a go.

**Comment by e0lithic** [comment]
Tried it , but the problem still persists.

**Comment by e0lithic** [comment]
Ok, I tried to create a virtualenv which is using the python from the root folders instead of pyenv's version. And in such a case, I am able to run the tool perfectly. It seems that when I using using a python installed by pyenv, that the library stops working. Is there a recommended way to use pyarmor with pyenv?

**Comment by jondy** [comment]
I know a little about pyenv. But if pyenv don't use CPython library, it's difficult to work with Pyarmor. If pyenv can be configured with CPython python library, it should be OK.

**Comment by e0lithic** [comment]
I have tried using `env PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 3.6.9`. Which should have ideally allowed .so to reference the cpython libraries as mentioned in the referenced issue above, but am still unable to get it running. Is there a workaround  or am I missing something?

**Comment by jondy** [comment]
What's the error message to run pyarmor?

**Comment by e0lithic** [comment]
## Command for build
`pyarmor obfuscate --advanced 1 --exact foo.py`

## Output on run
```
XXX lineno: 1, opcode: 228
Traceback (most recent call last):
  File "<foo.py>", line 3, in <module>
  File "<frozen foo>", line 1, in <module>
SystemError: unknown opcode
```

Tried this with both versions of python installs on pyenv
- `pyenv install 3.6.9`
- `env PYTHON_CONFIGURE_OPTS="--enable-shared"  pyenv install 3.6.9`

**Comment by jondy** [comment]
It seems that the python library used by pyenv isn't patched by pyarmor correctly. What the output with extra "-d" to run the obfuscated foo.py?
```
cd dist/
python -d foo.py
```

If it complains of the library patched failed, find the dynamic library used by pyenv, and upload it to the temporary site.

**Comment by e0lithic** [comment]
This is the output :
```
Initialize python patcher failed (1519)
Traceback (most recent call last):
  File "<foo.py>", line 3, in <module>
RuntimeError: The script is obfuscated in super mode but this library doesn't support
```
However, I have double checked that I have used advanced mode 1 (not 2)

**Comment by jondy** [comment]
The key problem is patch failed. I think pyenv uses different python dynamic library from the host python. You can start python interpreter in pyenv and check the dynamic libraries used by this process, I need check it.

**Comment by e0lithic** [comment]
Since I am not explicitly setting the LD_LIBRARY_PATH, I presume the libraries would be picked from the sys.path which : 
```
>>> import sys
>>> sys.path
['', '/home/e0lithic/.pyenv/versions/3.6.9/lib/python36.zip', '/home/e0lithic/.pyenv/versions/3.6.9/lib/python3.6', '/home/e0lithic/.pyenv/versions/3.6.9/lib/python3.6/lib-dynload', '/home/e0lithic/.local/lib/python3.6/site-packages', '/home/e0lithic/.pyenv/versions/3.6.9/lib/python3.6/site-packages']
```

Do you want me to upload the `/home/e0lithic/.pyenv/versions/3.6.9` folder?

**Comment by jondy** [comment]
No. What I care about only dynamic library. You can got it by this way:

1. Start python interpreter from pyenv
2. Get the pid of this interpreter by `ps`
3. Check the dynamic library used by this process
```
cat /proc/${pid}/maps | grep libpython*.so
```

Just upload this `.so` file.

**Comment by e0lithic** [comment]
Thanks for the detailed instruction. [Link](https://tmpfiles.org/download/50460/libpython3.6m.so.1.0)

**Comment by jondy** [comment]
@e0lithic  Could you upload it again? I try to download it just now, but it's gone.

**Comment by e0lithic** [comment]
@jondy I have uploaded it again. Here's the [link](https://tmpfiles.org/download/50636/libpython3.6m.so.1.0)

**Comment by jondy** [comment]
I have downloaded this one, It's OK, thanks.

**Comment by e0lithic** [comment]
@jondy Do you see a fix/workaround in near time frame for this. Or should I just temporary remove pyenv from my poetry env to make a couple of releases needed for my team.

**Comment by jondy** [comment]
@e0lithic I'll check it and feedback you tomorrow.

**Comment by jondy** [comment]
@e0lithic I build a temporary one [here](https://pyarmor.dashingsoft.com/downloads/temp/_pytransform.so), just replace the old one with it and check it work or not.

If it still doesn't work, run the obfuscated scripts with `-d`, and show the output message here.

**Comment by e0lithic** [comment]
Tried it . Got the following result :

```
Initialize python patcher failed (66889)
Traceback (most recent call last):
  File "<foo.py>", line 3, in <module>
RuntimeError: The script is obfuscated in advanced mode but the runtime library doesn't support
```

**Comment by jondy** [comment]
Try [this one](https://pyarmor.dashingsoft.com/downloads/temp/_pytransform.so)
md5sum: 0cc7b9f15dc7c4f0b442760f00a96f6f

**Comment by e0lithic** [comment]
Here is the output:
```
Initialize python patcher failed (66926)
Traceback (most recent call last):
  File "<foo.py>", line 3, in <module>
RuntimeError: The script is obfuscated in advanced mode but the runtime library doesn't support
```

**Comment by jondy** [comment]
Could you upload this file `/lib/x86_64-linux-gnu/libm.so.6` in your Ubuntu? It uses GLIBC-2.29 which missing in my ubuntu.

**Comment by e0lithic** [comment]
Heres the [link](https://tmpfiles.org/download/50847/libm.so.6)

**Comment by jondy** [comment]
Try [this one](https://pyarmor.dashingsoft.com/downloads/temp/_pytransform.so)
md5sum: 7bb8bbedc3f2bb2ea1e64ea4be5fa2d3

Hope it works

**Comment by e0lithic** [comment]
Got the following error nwo : 
```
Initialize python patcher failed (66932)
Traceback (most recent call last):
  File "<foo.py>", line 3, in <module>
RuntimeError: The script is obfuscated in advanced mode but the runtime library doesn't support
```

**Comment by jondy** [comment]
Try [this one](https://pyarmor.dashingsoft.com/downloads/temp/_pytransform.so)
md5sum: 9fe5857ffe3b88d34b73b1179bbd8771

**Comment by e0lithic** [comment]
It worked. :+1:

**Comment by jondy** [comment]
Nice, it'll be released with next version v6.2.2

**Comment by e0lithic** [comment]
Closing the issue as this has been fixed in the 6.2.2 release. Thanks for the great support.
